crates/core/src/events/service.rs

```diff
@@ -0,0 +1,328 @@
+use std::sync::Arc;
+
+use matrix::admin::resources::room::{MessagesParams, RoomEventFilter, RoomService};
+use matrix::client::resources::events::CreateEventResponse;
+use matrix::events::exports::ruma_common::{EventId, RoomId};
+use matrix::events::space::board::{
+    BoardPostEvent, BoardReplyEvent, SpaceReactionEventContent, Vote,
+};
+use matrix::events::{
+    exports::ruma_common::TransactionId,
+    space::board::{BoardPostEventContent, BoardReplyEventContent},
+};
+use matrix::events::{Raw, StateEventContent};
+use matrix::Client as MatrixAdminClient;
+use reqwest::Response;
+use serde::de::DeserializeOwned;
+use serde::Serialize;
+use serde_json::{from_str, json};
+
+use crate::events::error::BoardErrorCode;
+use crate::util::secret::Secret;
+use crate::{Error, Result};
+
+// TODO: find out how to serialize the full event
+// #[derive(Serialize)]
+pub struct GetPostResponse {
+    pub post: BoardPostEvent,
+    pub replies: Option<Vec<BoardReplyEvent>>,
+}
+
+pub struct EventsService {
+    admin: Arc<MatrixAdminClient>,
+}
+
+impl EventsService {
+    pub fn new(admin: Arc<MatrixAdminClient>) -> Self {
+        Self { admin }
+    }
+
+    async fn put_json(
+        &self,
+        path: impl AsRef<str>,
+        body: &impl Serialize,
+        access_token: Secret,
+    ) -> Result<Response> {
+        let mut tmp = (*self.admin).clone();
+        tmp.set_token(access_token.to_string()).map_err(|err| {
+            tracing::error!(?err, "Failed to set access token");
+            Error::Unknown
+        })?;
+
+        tmp.put_json(path, &body).await.map_err(|err| {
+            tracing::error!(?err, "Failed to make http request");
+            Error::Unknown
+        })
+    }
+
+    pub async fn send_post<S: AsRef<str>>(
+        &self,
+        content: Raw<BoardPostEventContent>,
+        board_id: S,
```

Comment:
```suggestion
        board_id: impl AsRef<str>,
```

Given that this is the single param that uses generic value